models/gif_cond_exp_multisynapse.cpp

```diff
@@ -0,0 +1,654 @@
+  B_.sys_.params = reinterpret_cast< void* >( this );
+}
+
+void
+nest::gif_cond_exp_multisynapse::calibrate()
+{
+
+  P_.receptor_types_.resize( P_.num_of_receptors_ );
+  for ( size_t i = 0; i < P_.num_of_receptors_; i++ )
+  {
+    P_.receptor_types_[ i ] = i + 1;
+  }
+
+  int state_size = 1 + ( State_::STATE_VEC_SIZE - 1 ) * P_.num_of_receptors_;
+
+  B_.spike_exc_.resize( P_.num_of_receptors_ );
+  B_.spike_inh_.resize( P_.num_of_receptors_ );
+  S_.y_.resize( state_size );
+
+  B_.sys_.dimension = state_size;
+
+  B_.logger_.init();
+
+  const double_t h = Time::get_resolution().get_ms();
+  V_.rng_ = kernel().rng_manager.get_rng( get_thread() );
+
+  V_.RefractoryCounts_ = Time( Time::ms( P_.t_ref_ ) ).get_steps();
+  // since t_ref_ >= 0, this can only fail in error
+  assert( V_.RefractoryCounts_ >= 0 );
+
+  // initializing adaptation (stc/sfa) variables
+  V_.P_sfa_.resize( P_.tau_sfa_.size(), 0.0 );
+  V_.P_stc_.resize( P_.tau_stc_.size(), 0.0 );
+
+  for ( size_t i = 0; i < P_.tau_sfa_.size(); i++ )
+  {
+    V_.P_sfa_[ i ] = std::exp( -h / P_.tau_sfa_[ i ] );
+  }
+  S_.sfa_elems_.resize( P_.tau_sfa_.size(), 0.0 );
+
+  for ( size_t i = 0; i < P_.tau_stc_.size(); i++ )
+  {
+    V_.P_stc_[ i ] = std::exp( -h / P_.tau_stc_[ i ] );
+  }
+  S_.stc_elems_.resize( P_.tau_stc_.size(), 0.0 );
+}
+
+/* ----------------------------------------------------------------
+ * Update and spike handling functions
+ */
+
+void
+nest::gif_cond_exp_multisynapse::update( Time const& origin,
+  const long_t from,
+  const long_t to )
+{
+
+  assert(
+    to >= 0 && ( delay ) from < kernel().connection_manager.get_min_delay() );
+  assert( from < to );
+
+  for ( long_t lag = from; lag < to; ++lag )
+  {
+
+    // exponential decaying stc and sfa elements
+    S_.stc_ = 0.0;
+    for ( size_t i = 0; i < S_.stc_elems_.size(); i++ )
+    {
+      S_.stc_ += S_.stc_elems_[ i ];
+      S_.stc_elems_[ i ] = V_.P_stc_[ i ] * S_.stc_elems_[ i ];
+    }
+
+    S_.sfa_ = P_.V_T_star_;
+    for ( size_t i = 0; i < S_.sfa_elems_.size(); i++ )
+    {
+      S_.sfa_ += S_.sfa_elems_[ i ];
+      S_.sfa_elems_[ i ] = V_.P_sfa_[ i ] * S_.sfa_elems_[ i ];
+    }
+
+    // numerical integration with adaptive step size control:
+    // ------------------------------------------------------
+    // gsl_odeiv_evolve_apply performs only a single numerical
+    // integration step, starting from t and bounded by step;
+    // the while-loop ensures integration over the whole simulation
+    // step (0, step] if more than one integration step is needed due
+    // to a small integration step size;
+    // note that (t+IntegrationStep > step) leads to integration over
+    // (t, step] and afterwards setting t to step, but it does not
+    // enforce setting IntegrationStep to step-t; this is of advantage
+    // for a consistent and efficient integration across subsequent
+    // simulation intervals
+
+    double t = 0.0;
+
+    double_t* state_y_ = &S_.y_[ 0 ];
```

Comment:
This is unsafe, since it assumes that the data of a `std::vector<double>` has the same layout as a C-style array of doubles. While this tends to hold true, it is not, I believe, required by the C++ standard and we should not rely on it. Rather, `S_.y_` should be a C-style array instead of a C++ vector.